include/xtensor/xhistogram.hpp

```diff
@@ -57,46 +57,94 @@ namespace xt
     inline auto histogram(E1&& data, E2&& bin_edges, E3&& weights, bool density = false)
     {
-        using size_type = common_size_type_t<std::decay_t<E1>, std::decay_t<E2>, std::decay_t<E3>>;
-        using value_type = typename std::decay_t<E3>::value_type;
-
-        XTENSOR_ASSERT(data.dimension() == 1);
-        XTENSOR_ASSERT(weights.dimension() == 1);
-        XTENSOR_ASSERT(bin_edges.dimension() == 1);
-        XTENSOR_ASSERT(weights.size() == data.size());
-        XTENSOR_ASSERT(bin_edges.size() >= 2);
-        XTENSOR_ASSERT(std::is_sorted(bin_edges.cbegin(), bin_edges.cend()));
-        XTENSOR_ASSERT(xt::amin(data)[0] >= bin_edges[0]);
-        XTENSOR_ASSERT(xt::amax(data)[0] <= bin_edges[bin_edges.size() - 1]);
-
-        xt::xtensor<value_type, 1> count = xt::zeros<value_type>({ bin_edges.size() - 1 });
-
-        auto sorter = xt::argsort(data);
-
-        size_type ibin = 0;
+        return histogramImp(std::forward<E1>(data),
+                            std::forward<E2>(data),
+                            std::forward<E3>(weights),
+                            density,
+                            false);
+    }
 
-        for (auto& idx : sorter)
+    namespace detail
+    {
+        /**
+         * @ingroup histogram
+         * @brief Compute the histogram of a set of data.
+         *
+         * @param data The data.
+         * @param bin_edges The bin-edges. It has to be 1-dimensional and monotonic.
+         * @param weights Weight factors corresponding to each data-point.
+         * @param density If true the resulting integral is normalized to 1. [default: false]
+         * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+         */
+        template <class R = double, class E1, class E2, class E3>
+        inline auto histogramImp(E1&& data, E2&& bin_edges, E3&& weights, bool density = false, bool equal_bins=false)
         {
-            while (data[idx] >= bin_edges[ibin + 1] && ibin < bin_edges.size() - 2)
+            using size_type = common_size_type_t<std::decay_t<E1>, std::decay_t<E2>, std::decay_t<E3>>;
+            using value_type = typename std::decay_t<E3>::value_type;
+
+            XTENSOR_ASSERT(data.dimension() == 1);
+            XTENSOR_ASSERT(weights.dimension() == 1);
+            XTENSOR_ASSERT(bin_edges.dimension() == 1);
+            XTENSOR_ASSERT(weights.size() == data.size());
+            XTENSOR_ASSERT(bin_edges.size() >= 2);
+            XTENSOR_ASSERT(std::is_sorted(bin_edges.cbegin(), bin_edges.cend()));
+            XTENSOR_ASSERT(xt::amin(data)[0] >= bin_edges[0]);
+            XTENSOR_ASSERT(xt::amax(data)[0] <= bin_edges[bin_edges.size() - 1]);
+
+            size_t n_bins = bin_edges.size() - 1;
+            xt::xtensor<value_type, 1> count = xt::zeros<value_type>({ n_bins });
+
+            if (equal_bins)
             {
-                ++ibin;
+                std::array<double, 2> bounds = xt::minmax(data)();
```

Comment:
`xt::minmax(a)` returns `std::array<type, 2>` where `type` is `typename std::decay_t<decltype(a)>::value_type`. Since `std::array` does not provide any conversion constructor / operator, if the `value_type` of `data` is not double, this will not compile.

Reply:
Thanks for the review and the explanation! I always thought there could be some difference between `xtensor` and `xarray` :-(